### ISSUE TITLE
Improvements to status server

### ISSUE THREAD

**Issue by bwhm** [Joystream/status-endpoint-joystream#12]
Some of the items below will require changes on the [website](https://github.com/Joystream/joystream-org) as well, but afaict, all the work on the status-server can be made first.

## Burning of Tokens
The burning of received tokens doesn't always work. It has happened on at least two occasions, where I have made them manually. 

Example at block 2358370, where the tokens were received, and parsed in `out.log`, but the subsequent entries doesn't show any burns taking place, and exchanges.json shows:

```
    {
      "amount": 2000000,
      "tokensRecievedAtBlock": 2358370,
      "finalStatus": "Invalid"
    },
```

This can, and will, happen from time to time, so I don't think it's worth trying multiple times if the burn fails. We regularly check and process the exchanges anyway.

The problem is that when I have burned them manually later, they don't get registered as burned in the log/db, and therefore, not [here](https://status.joystream.org/status). My suggestion would be to:

1. Keep the processing and initiating as before, but decouple the "burn" from the exchange.
2. Instead, also subscribe to extrinsics, and if the exchange address is the signer, use this to log new burns. This will catch all burns, regardless of manual or not.
3. Optionally, don't distinguish between `totalIssuance` and `actualIssuance`. If a transaction isn't burned automatically, we'll deal with it manually, and the burn will get registered as it should. The small offset in exchange rate is ok.

If you have a better idea, feel free...

## Error log
Could you add the timestamp here as well? Won't be from the chain ofc, but it's hard to parse when something happened...

## Changing exchanges.json
There are a couple of things here, not all are that critical.

### Total `amountUSD`
Add an entry that sums and displays the USD value of all exchanges. 

### Total `addedToFiatPool`
This may either be something we add manually, or, if below is possible, even better. 

---

This may be more time consuming than it's worth, but would be nice:

## Command line writing
Is it possible to add commands that are processed? Currently, we have to stop the server and manually edit the exchange.json file. This leaves room for human error, and causes downtime on the whole status server. It also makes it impossible to use a load balancer, or some backup server, that is perfectly in sync (this would be tricky anyway, but possible)

1. Change the `sizeDollarPool`
If I wanted to change the size of the fiat, just make a command like:
`changeDollarPool <amount> [block]`
So if I wanted to add $100 immidiately:
`changeDollarPool 100`

The blockheight arg would be even nicer, but probably more difficult, so a lot less important...

2. Status of `exchanges`
Instead of having the delete or comment out exchanges, add a `status` key for each exchange that defaults to `unprocessed` or just `0`
When it's paid, I make a command like:
`changeStatus <index> <status>`
So when exchange `13` is paid:
`changeStatus 13 1`
...and the website won't list it anymore.

And alternative could be to add a remove command...

**Comment by Lezek123** [comment]
> (...) The problem is that when I have burned them manually later, they don't get registered as burned in the log/db

Actually the sum of burned tokens (`burned` field from https://status.joystream.org/status) seems to be updated on each _Exchange_, regardless of the status of the actual "real burning" transaction, so:

1. I start a fresh status server with `burned = 0`
2. I send `100 JOY` to burning address
3. `balances.transfer` event is parsed by the status server - `burned` gets updated to `100` (the tokens are not yet actually burned, but it's already assumed that they will be)
4. Meanwhile the "real burn" transaction is initialized, but the result doesn't affect the value of `burned` it only add a new entry into `burns` (so the status of the automatic "real burn" transactions can be tracked and optionally manual burning can be performed if required, but this shouldn't cause the need to update the `exchanges.json`)

> Optionally, don't distinguish between totalIssuance and actualIssuance. If a transaction isn't burned automatically, we'll deal with it manually, and the burn will get registered as it should. The small offset in exchange rate is ok.

The `totalIssuance` and `actualIssuance` isn't taking the value of `burned` or `burns` into account (so those don't really affect the exchange rate). The formula to calculate `actualIssuance` is just `totalIssuance - currentBurningAddressBalance`.

`burned` is only a value that tracks the total amount of tokens burned from the beginning of parsing (which should just effectively be the sum of all tokens sent to burning address ever, calculated based on all the parsed `balances.transfer` events data). It is a value exposed by the status server for infromative purposes, but it's not used internally by it to perform any calculations.

**Comment by bwhm** [comment]
> Actually the sum of burned tokens (`burned` field from https://status.joystream.org/status) seems to be updated on each _Exchange_, regardless of the status of the actual "real burning" transaction (whether manual or automatic), so:

Yes, the `burns` stores all `events` where the address receives tokens. However, when you send to yourself (which I do when burning "manually", it is not an on chain `event`, **only** an `extrinsics`, [ref](https://testnet.joystream.org/#/explorer/query/2385655). So I think I understood the process.

>     1. I start a fresh status server with `burned = 0`
> 
>     2. I send `100 JOY` to burning address
> 
>     3. `balances.transfer` event is parsed by the status server - `burned` gets updated to `100` (the tokens are not yet actually burned, but it's already assumed that they will be)
> 
>     4. Meanwhile the "real burn" transaction is initialized, but the result doesn't affect the value of `burned` it only add a new entry into `burns` (so the status of the automatic "real burn" transactions can be tracked and optionally manual burning can be performed if required, but this shouldn't cause the need to update the `exchanges.json`)

Yes, this is what I've observed as well, but I **think** the `tokensBurned` doesn't get registered unless the burn gets `Finalized`.


> The `totalIssuance` and `actualIssuance` isn't taking the value of `burned` or `burns` into account (so those don't really affect the exchange rate). The formula to calculate `actualIssuance` is just `totalIssuance - burningAddressBalance` (at given block).

> `burned` is only a value that tracks the total amount of tokens recieved by the burning address from the beginning of parsing (it sums it up based on the parsed `balances.transfer` events data). It is a value exposed by the status server for infromative purposes, but it's not used internally by it to perform any calculations.

That may be true, but I do have a discrepancy between the sum of `burns` and the `tokensBurned` atm. It may be human error on my part though.

Although it's not used by the status server itself, it would be useful internally!

**Comment by bwhm** [comment]
Adding a small feature request here:

`yarn increasePool <amount> [blockHeight] [comment]

In addition to the current `exchanges.json`, add `poolChangeHistory`.
For every top-up (`increasePool`) or entry in `exchanges`:

```
"poolChangeHistory": [
   {
	blockheight: i,
	timeStamp: timeStampOf[i],
	issuance: issuance.at[i],
	change: amountAddedOrRemoved,       // +-increasePool or -exchangeValue
	valueAfter: poolChangeHistory[poolChangeHistory.length-1].valueAfter - change,
	rateAfter: valueAfter/issuance,
	reason: “reason”,                                   // increasePool "comment", or "exchange"
   }
]

**Comment by bwhm** [comment]
Closed with #13
